src/main/java/org/jenkinsci/plugins/buildnameupdater/BuildNameUpdater.java

```diff
@@ -0,0 +1,243 @@
+package org.jenkinsci.plugins.buildnameupdater;
+
+import hudson.Extension;
+import hudson.FilePath;
+import hudson.Launcher;
+import hudson.model.AbstractBuild;
+import hudson.model.AbstractProject;
+import hudson.model.BuildListener;
+import hudson.model.FreeStyleProject;
+import hudson.remoting.VirtualChannel;
+import hudson.tasks.BuildStepDescriptor;
+import hudson.tasks.Builder;
+import hudson.util.FormValidation;
+import org.jenkinsci.plugins.tokenmacro.MacroEvaluationException;
+import org.jenkinsci.plugins.tokenmacro.TokenMacro;
+import org.kohsuke.stapler.DataBoundConstructor;
+import org.kohsuke.stapler.QueryParameter;
+
+import javax.servlet.ServletException;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+
+/**
+ *
+ * This plugin replace the build name with the first line from a file on a slave.
+ *
+ * @author Lev Mishin
+ */
+public class BuildNameUpdater extends Builder {
+
+    private final String buildName;
+    private final String macroTemplate;
+    private final boolean fromFile;
+    private final boolean fromMacro;
+    private final boolean macroFirst;
+
+    // Fields in config.jelly must match the parameter names in the "DataBoundConstructor"
+    @DataBoundConstructor
+    public BuildNameUpdater(BuildNameFromFile nameFromFile, BuildNameFromMacro nameFromMacro) {
```

Comment:
I doubt you really need these classes. You could use "optionalBlock" instead of the property and get the same result without new classes. Example: https://github.com/jenkinsci/walldisplay-plugin/blob/ebeeb14f36573a702a7758280a548da46f9648c6/src/main/resources/de/pellepelster/jenkins/walldisplay/WallDisplayJobProperty/config.jelly